synaptic_reconstruction/tools/vesicle_pool_widget.py

```diff
@@ -0,0 +1,219 @@
+            show_info("INFO: Please choose a segmentation.")
+            return
+        if self.query_param.text() == "":
+            show_info("INFO: Please enter a query string.")
+            return
+        # resolve string
+        query = self.query_param.text()
+
+        if self.new_layer_name_param.text() == "":
+            show_info("INFO: Please enter a new layer name.")
+            return
+        new_layer_name = self.new_layer_name_param.text()
+        if self.pooled_group_name_param.text() == "":
+            show_info("INFO: Please enter a pooled group name.")
+            return
+        pooled_group_name = self.pooled_group_name_param.text()
+
+        if distances is None:
+            show_info("INFO: Distances layer could not be found or has no values.")
+            return
+        self._compute_vesicle_pool(segmentation, distances, morphology, new_layer_name, pooled_group_name, query)
+
+    def _compute_vesicle_pool(self, segmentation, distances, morphology, new_layer_name, pooled_group_name, query):
+        """
+        Compute a vesicle pool based on the provided query parameters.
+
+        Args:
+            segmentation (array): Segmentation data (e.g., labeled regions).
+            distances (dict): Properties from the distances layer.
+            morphology (dict): Properties from the morphology layer.
+            new_layer_name (str): Name for the new layer to be created.
+            pooled_group_name (str): Name for the pooled group to be assigned.
+            query (dict): Query parameters.
+        """
+
+        distances_ids = distances.get("id", [])
+        morphology_ids = morphology.get("label_id", [])
+
+        # Check if IDs are identical
+        if set(distances_ids) != set(morphology_ids):
+            show_info("ERROR: The IDs in distances and morphology are not identical.")
+            return
+
+        distances = pd.DataFrame(distances)
+        morphology = pd.DataFrame(morphology)
+
+        # Merge dataframes on the 'id' column
+        merged_df = morphology.merge(distances, left_on="label_id", right_on="id", suffixes=("_morph", "_dist"))
+
+        # Apply the query string to filter the data
+        filtered_df = self._parse_query(query, merged_df)
+
+        # Extract valid vesicle IDs
+        valid_vesicle_ids = filtered_df["label_id"].tolist()
+
+        new_layer_data = np.zeros(segmentation.shape, dtype=np.uint8)
+        pool_id = 1
+        layer = None
+
+        # check if group already exists 
+        if new_layer_name in self.viewer.layers:
+            layer = self.viewer.layers[new_layer_name]
+            if pooled_group_name not in layer.properties["pool"]:
+                new_layer_data = layer.data
+                pool_id = len(np.unique(layer.properties["pool"])) + 1
+        # compute vesicles with new pool_id and properties
+        for vesicle_id in valid_vesicle_ids:
+            new_layer_data[segmentation == vesicle_id] = pool_id
```

Comment:
There are a few problems here:
- You set the value in the layer to the `pool_id`. Instead we should keep the vesicle ids intact, but just filter out the vesicles that are not part of any pool.
- The whole pool assignment and property logic could be simplified a lot, see the general comment for how.
- The way you do this is in general inefficient. Instead of solving this with a loop you can for example do
```python
filtered_vesicle_mask = ~np.isin(segmentation, valid_vesicle_ids)
new_layer_data[filtered_vesicle_mask] = 0
```